### PR TITLE
BFD-2244: Enable SLO alerts for existing server SLOs and other alarms

### PR DESCRIPTION
**Pull request by malessi**
**JIRA Ticket:**
[BFD-2244](https://jira.cms.gov/browse/BFD-2244)

**User Story or Bug Summary:**

As a BFD Engineer, I want to know when the BFD Server application is performing outside of our
agreed upon SLOs and to be able to immediately address any problems. Additionally, I want to be sure
that when a BFD Engineer is paged for an SLO exceeding its "alert" threshold that it is not a
false-positive.

---

### What Does This PR Do?

This PR removes the temporary SNS topics introduced for testing CloudWatch Alarms for BFD Server
SLOs and replaces them with permanent topics that route corresponding to the alarm's established
environment and implicit severity. The following SNS topics have been added to the `stateful`
module, with `bfd-ENV-cloudwatch-alarms-slack-bfd-test` essentially replacing the existing,
temporary SNS topics used by most alarms up to this point:

| SNS Topic | Destination |
| --- | --- |
| `bfd-ENV-cloudwatch-alarms-slack-bfd-test` | The #bfd-test Slack channel |
| `bfd-ENV-cloudwatch-alarms-slack-bfd-warnings` | The #bfd-warnings Slack channel |
| `bfd-ENV-cloudwatch-alarms-slack-bfd-alerts` | The #bfd-alerts Slack channel |

Plainly, here is where each type of alarm is routed to per-environment:

| Alarm Type or Alarm Group | `prod` | `prod-sbx` | `test` |
| --- | --- | --- | --- |
| BFD Server **_Alert_** SLO Alarms | -On `ALARM`: To VictorOps, which will page the on-call engineer and post to #bfd-alerts<br>-On `OK`: To the `bfd-cloudwatch-prod-ok` SNS Topic which routes to VictorOps | -On `ALARM`: To the #bfd-test Slack channel<br>-On `OK`: Nowhere | -On `ALARM`: To the #bfd-test Slack channel<br>-On `OK`: Nowhere |
| BFD Server **_Warning_** SLO Alarms | -On `ALARM`: To the #bfd-warnings Slack channel<br>-On `OK`: Nowhere | -On `ALARM`: To the #bfd-test Slack channel<br>-On `OK`: Nowhere | -On `ALARM`: To the #bfd-test Slack channel<br>-On `OK`: Nowhere |
| BFD Server Log Availability Alarm | -On `ALARM`: To VictorOps, which will page the on-call engineer and post to #bfd-alerts<br>-On `OK`: To the `bfd-cloudwatch-prod-ok` SNS Topic which routes to VictorOps | -On `ALARM`: To VictorOps, which will page the on-call engineer and post to #bfd-alerts<br>-On `OK`: To the `bfd-cloudwatch-prod-sbx-ok` SNS Topic which routes to VictorOps | -On `ALARM`: Nowhere (_`test` is expected to have no activity at some times)_<br>-On `OK`: Nowhere |
| BFD Server Query Logging Listener | -On `ALARM`: To the #bfd-warnings Slack channel<br>-On `OK`: Nowhere | -On `ALARM`: To the #bfd-warnings Slack channel<br>-On `OK`: Nowhere | -On `ALARM`: To the #bfd-test Slack channel<br>-On `OK`: Nowhere |
| BFD Server Disk Usage Alarms | -On `ALARM`: To VictorOps, which will page the on-call engineer and post to #bfd-alerts<br>-On `OK`: Nowhere | -On `ALARM`: To VictorOps, which will page the on-call engineer and post to #bfd-alerts<br>-On `OK`: Nowhere | -On `ALARM`: To the #bfd-test Slack channel<br>-On `OK`: Nowhere |
| BFD Pipeline Log Availability Alarm | -On `ALARM`: To VictorOps, which will page the on-call engineer and post to #bfd-alerts<br>-On `OK`: Nowhere | -On `ALARM`: To the #bfd-alerts Slack channel<br>-On `OK`: Nowhere | -On `ALARM`: To the #bfd-test Slack channel<br>-On `OK`: Nowhere |

I have done my best, where not explicitly defined by BFD-2244, to choose destinations for alerts
that are appropriate based upon the environment and the alarm's severity. _I am looking for_
_feedback on these choices, and am open to changing them based on what the team feels is most_
_appropriate for each type of alarm_.

Additionally, refactoring has been done to how the `bfd_server_slo_alarms`, `bfd_server_log_alarms`,
and `disk_usage_alarms` modules resolve their alarms' respective "action" and "ok" SNS topics.
Previously, these modules received the SNS topics their alarms used from their parent module
(`stateless`) through Terraform variables; however, this behavior was difficult to reconcile with
ephemeral environments and the fact that the alarms within each module each had different severities
of alarms and therefore needed to be routed to different destinations on a per-module basis. Thus,
the per-environment resolution of SNS topics has been moved internally to each module, with each
module having override variables defined for each type of SNS topic destination. Additionally, with
this change, ephemeral environments are now handled much more gracefully as alarm actions will
simply be omitted if the environment has no defined SNS destination for a particular type of alarm.

Finally, the `cw_alarms_slack_notifier` module has been updated to create Lambdas for the 3 new SNS
topics listed above, with each routing Alarm notifications to their corresponding Slack channels.

The changes to the `stateful` module will be applied once this PR is merged. The changes in this PR
have been verified to be correct through `terraform plan`ning each module individually via the
`stateless` module and manually verifying that alarm destinations were set to the expected SNS
topic.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

- Are the choices for alarm destinations appropriate for each type of alarm?
- Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

- If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

  - Does this PR add any new software dependencies?
    - [ ] Yes
    - [x] No
  - Does this PR modify or invalidate any of our security controls?
    - [ ] Yes
    - [x] No
  - Does this PR store or transmit data that was not stored or transmitted before?
    - [ ] Yes
    - [x] No

- If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
  - Do you think this PR requires additional review of its security implications for other reasons?
    - [ ] Yes
    - [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

- N/A

### Submitter Checklist

I have gone through and verified that...:

- [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
- [x] This PR is reasonably limited in scope, to help ensure that:
  1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
  2. There isn't too much of a burden on reviewers.
  3. Any problems it causes have a small "blast radius".
  4. It'll be easier to rollback if that becomes necessary.
- [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
- [x] The data dictionary has been updated with any field mapping changes, if any were made.
- [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
- [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
- [x] Reviews are requested from both:
  - At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
  - Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
- [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
  - Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: ops/terraform/modules/resources/bfd_server_slo_alarms/main.tf
Comment (mjburling) [inline ops/terraform/modules/resources/bfd_server_slo_alarms/main.tf:46]: Interesting. I've never had a case for using `try` in vanilla terraform. Hmm.

Comment (malessi) [reply ops/terraform/modules/resources/bfd_server_slo_alarms/main.tf:46]: I'd be interested to know if there's a more "Terraformic" (i.e. _Pythonic_) way of doing the operations that are being done here. Basically, `coalesce` will error if all of its parameters evaluate to `null`, so I need to wrap that in `try` in order to ensure that if they're all `null` `null` is returned instead of an error. The second case usage of `try` here is _basically_ the equivalent of the `?.[n]` operator or wrapping the array with a `if array.length > 0` or something similar. I'm not sure if there's the equivalent of `map`'s `lookup` for arrays in Terraform, but if I'm wrong then that might be a better approach.

File: ops/terraform/modules/resources/bfd_server_log_alarms/main.tf
Comment (lsmitchell) [inline ops/terraform/modules/resources/bfd_server_log_alarms/main.tf:20]: Is this true? I can't recall a time test wasn't running (save for a bad deployment or something)

**Review by lsmitchell — COMMENTED** [review]
Mostly agree with the decisions re: routing alerts and alarms. My one thought was that the prod-sbx "downgraded" items might want to be the same as prod. Thinking is that we treat prod-sbx as a prod environment for most other things, and partners do somewhat depend on the availability of that environment as well. I could also see the argument that someone doesnt necessarily need to get out of bed to fix prod-sbx if something were to happen in the middle of the night, so either way is fine by me, but thought I'd give that other perspective.

File: ops/terraform/modules/resources/bfd_server_log_alarms/main.tf
Comment (malessi) [reply ops/terraform/modules/resources/bfd_server_log_alarms/main.tf:20]: This isn't necessarily true just yet, but there was some discussion recently about the _possibility_ of shutting `test` down after hours or something similar and I believe this particular log availability alarm was discussed as potentially causing some noise if that were ever to happen. So, this is more preemptive than anything as we could _technically_ turn the alarm on right now without there being any false positives since `test` is always running.

**Comment by malessi** [comment]
> Mostly agree with the decisions re: routing alerts and alarms. My one thought was that the prod-sbx "downgraded" items might want to be the same as prod. Thinking is that we treat prod-sbx as a prod environment for most other things, and partners do somewhat depend on the availability of that environment as well. I could also see the argument that someone doesnt necessarily need to get out of bed to fix prod-sbx if something were to happen in the middle of the night, so either way is fine by me, but thought I'd give that other perspective.

I definitely understand that -- the BFD Server Log Availability alarm for `prod-sbx` does actually send alerts to VictorOps, so we're already treating `prod-sbx` as an environment worthy of waking someone up in that case. 

Maybe this is something we talk about more as a team before this gets merged? @mjburling thoughts?

**Review by lsmitchell — APPROVED** [review]

File: ops/terraform/modules/resources/bfd_server_slo_alarms/main.tf
Comment (mjburling) [reply ops/terraform/modules/resources/bfd_server_slo_alarms/main.tf]: The future's bright. Thanks for the thorough explanation!

**Review by mjburling — APPROVED** [review]

**Review by lsmitchell — APPROVED** [review]

**Comment by mjburling** [comment]
> Maybe this is something we talk about more as a team before this gets merged? @mjburling thoughts?

As discussed in standup on 2023-02-20 (and as you've already gone ahead and implemented here), the team agrees that a subset of `prod-sbx` alarms will route as the `prod` alarms do.

**Review by mjburling — APPROVED** [review]

**Comment by lsmitchell** [comment]
Approved, assuming that synk thing isn't something serious.

**Pull request merged**
